core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedService.java

```diff
@@ -420,63 +419,80 @@ private Object invoke(ServiceRequestContext ctx, HttpRequest req, AggregatedResu
-                                    : ResponseHeaders.builder().add(headers));
-            newTrailers = trailers;
-        }
-
+    private HttpResponse convertResponse(ServiceRequestContext ctx, @Nullable Object result) {
         if (result instanceof HttpResponse) {
             return (HttpResponse) result;
         }
         if (result instanceof AggregatedHttpResponse) {
             return ((AggregatedHttpResponse) result).toHttpResponse();
         }
+
+        final ResponseHeaders headers;
+        final HttpHeaders trailers;
+
+        if (result instanceof HttpResult) {
+            final HttpResult<?> httpResult = (HttpResult<?>) result;
+            headers = buildResponseHeaders(ctx, httpResult.headers());
+            result = httpResult.content();
+            trailers = httpResult.trailers();
+        } else {
+            headers = buildResponseHeaders(ctx);
+            trailers = HttpHeaders.of();
+        }
+
+        return convertResponseInternal(ctx, headers, result, trailers);
+    }
+
+    private HttpResponse convertResponseInternal(ServiceRequestContext ctx,
+                                                 ResponseHeaders headers,
+                                                 @Nullable Object result,
+                                                 HttpHeaders trailers) {
         if (result instanceof CompletionStage) {
             final CompletionStage<?> future = (CompletionStage<?>) result;
-            return HttpResponse.from(future.thenApply(object -> convertResponse(ctx, newHeaders, object,
-                                                                                newTrailers)));
+            return HttpResponse.from(
+                    future.thenApply(object -> convertResponseInternal(ctx, headers, object, trailers)));
         }
 
         try (SafeCloseable ignored = ctx.push()) {
-            return responseConverter.convertResponse(ctx, newHeaders, result, newTrailers);
+            return responseConverter.convertResponse(ctx, headers, result, trailers);
         } catch (Exception cause) {
             return HttpResponse.ofFailure(cause);
         }
     }
 
-    private static ResponseHeadersBuilder addNegotiatedResponseMediaType(ServiceRequestContext ctx,
-                                                                         HttpHeaders headers) {
+    private ResponseHeaders buildResponseHeaders(ServiceRequestContext ctx, HttpHeaders customHeaders) {
+        final ResponseHeadersBuilder builder;
 
-        final MediaType negotiatedResponseMediaType = ctx.negotiatedResponseMediaType();
-        if (negotiatedResponseMediaType == null || headers.contentType() != null) {
-            // Do not overwrite 'content-type'.
-            return ResponseHeaders.builder()
-                                  .add(headers);
+        // Prefer ResponseHeaders#toBuilder because builder#add is an expensive operation.
```

Comment:
Let's use `builder#add(Iterable)` because it may confuse others. Or we can remove this sentence completely.